python/lsst/pipe/base/pipeTools.py

```diff
@@ -153,14 +153,20 @@ def orderPipeline(pipeline: List[TaskDef]) -> List[TaskDef]:
     outputs = {}  # maps task index to its output DatasetType names
     allInputs = set()  # all inputs of all tasks
     allOutputs = set()  # all outputs of all tasks
+    dsTypeTaskLabels: dict[str, str] = {}  # maps DatasetType name to the label of its parent task
     for idx, taskDef in enumerate(pipeline):
         # task outputs
         dsMap = {name: getattr(taskDef.connections, name) for name in taskDef.connections.outputs}
         for dsTypeDescr in dsMap.values():
             if dsTypeDescr.name in allOutputs:
                 raise DuplicateOutputError(
-                    "DatasetType `{}' appears more than once as output".format(dsTypeDescr.name)
+                    "DatasetType `{}' in task `{}' already appears as an output in task `{}'.".format(
+                        dsTypeDescr.name,
+                        taskDef.label,
+                        dsTypeTaskLabels[dsTypeDescr.name],
+                    )
                 )
+            dsTypeTaskLabels[dsTypeDescr.name] = taskDef.label
```

Comment:
The fact that the test is against `allOutputs` while the insertion is against `dsTypeTaskLabels` is a bit awkward, logically. Is it already guaranteed that a task can't have two connections with the same name?

Reply:
A single task cannot output the same dataset type twice. The important insertion is `dsTypeDescr.name` into `outputs`, which then gets subsumed into `allOutputs` to facilitate future checks. The insertion you've highlighted here is merely a book-keeping exercise that keeps a record of the prior tasks which output a given dataset type.